alf/algorithms/rl_algorithm.py

```diff
@@ -265,11 +366,47 @@ def predict(self, time_step: ActionTimeStep, state=None):
                 `action_distribution_spec`
               policy_step.state should be consistent with `predict_state_spec`
         """
-        policy_step = self.rollout(time_step, state)
+        if self.__class__.predict_action_distribution != RLAlgorithm.predict_action_distribution:
+            # predict_action_distribution is overridden, use it.
+            policy_step = self.predict_action_distribution(time_step, state)
+            action = common.sample_action_distribution(policy_step.action)
+            return policy_step._replace(action=action)
+        else:
+            # predict_action_distribution is not overridden. Use rollout()
+            policy_step = self.rollout(time_step, state)
+            return policy_step._replace(info=())
+
+    # Subclass may override predict_action_distribution()
+    def predict_action_distribution(self,
+                                    time_step: ActionTimeStep,
+                                    state=None):
+        """Predict for one step of observation.
+
+        This only used for evaluation. So it only need to perform compuations
+        for generating action distribution.
+
+        Args:
+            time_step (ActionTimeStep): Current observation and other inputs
+                for computing action.
+            state (nested Tensor): should be consistent with predict_state_spec
+        Returns:
+            policy_step (PolicyStep):
+              policy_step.action is nested Distribution which consistent with
+                `action_distribution_spec`
+              policy_step.state should be consistent with `predict_state_spec`
+        """
+        policy_step = self.rollout_action_distribution(time_step, state)
         return policy_step._replace(info=())
 
+    ON_POLICY_TRAINING = 0
+    OFF_POLICY_TRAINING = 1
+    ROLLOUT = 2
+
     @abstractmethod
```

Comment:
can we make deterministic policy algorithm override `predict_action_distribution` or `rollout_action_distribution` and return deterministic distribution if necessary , and all subclasses do not override `rollout` and `predict`, otherwise, it will make the code very confusing

Reply:
As an interface, `rollout` and `predict` are better because they are more general - the policy only output an action. `predict_action_distribution` and `rollout_action_distribution` are only useful for `greedy_predict`. Perhaps we can remove `predict_action_distribution`, `rollout_distribution` and `greedy_predict`, and change `predict` to include an additional argument `epsilon_greedy`?
```
@abstractmethod
def predict(self, time_step, state, epsilon_greedy)
```
And do not provide default implementation for predict. Subclass has to implement  it.  We can provide an utility function for generating epsilon greedy action from a distribution. This way the interface will be much cleaner, but at the expense of more code in the subclass.

I was thinking about this way but eventually changed to the current form. See https://github.com/HorizonRobotics/alf/blob/f18d45ed271bca4ba65353d9f297605d07259394/alf/utils/common.py#L691. 

What do you think @witwolf @hnyu?

Reply:
Now I removed rollout_action_distribution and predict_action_distribution. The subclasses do not seem to need much more lines.